File: src/pattern/visitor/java/ListVisitor.java
package pattern.visitor.java;


import java.util.Iterator;

public class ListVisitor extends Visitor {
    String currentDir ="" ;
    public void visit(File file){
        System.out.println(this.currentDir+"/"+file);
    }

    @Override
    void visit(Directory dir) {
        System.out.println(currentDir+"/"+dir);
        String savedir = currentDir;
        currentDir = currentDir + "/" + dir.getName();

        Iterator it = dir.iterator();
        while (it.hasNext()){
            Entry entry = (Entry)it.next();
            entry.accept(this);
        }

        currentDir = savedir;
    }
}

File: src/pattern/visitor/kotlin/ListVisitor.kt
package pattern.visitor.kotlin

class ListVisitor:Visitor(){
    var currentDir = ""
    override fun visit(file: File) {
        println("${this.currentDir}/${file}")
    }

    override fun visit(dir: Directory) {
        println("${this.currentDir}/${dir}")
        var savedDir = this.currentDir
        this.currentDir = this.currentDir + "/" +dir.getName()
        val it = dir.iterator()
        while (it.hasNext()){
            val entry  = it.next()
            entry.accept(this)
        }

        this.currentDir = savedDir
    }
}